include/engine/routing_algorithms/routing_base.hpp

```diff
@@ -44,71 +44,131 @@ bool needsLoopBackwards(const PhantomNode &source_phantom, const PhantomNode &ta
 bool needsLoopForward(const PhantomNodes &phantoms);
 bool needsLoopBackwards(const PhantomNodes &phantoms);
 
-template <typename Heap>
-void insertNodesInHeaps(Heap &forward_heap, Heap &reverse_heap, const PhantomNodes &nodes)
+namespace detail
+{
+template <typename Algorithm>
+void insertSourceInHeap(typename SearchEngineData<Algorithm>::ManyToManyQueryHeap &heap, const PhantomNode &phantom_node)
```

Comment:
Hm, why only template the algorithm and not simply the heap? Does it make sense to restricti it to a specific heap impl.?

Reply:
We need this in order to provide overloads for the normal QueryHeap and ManyToMany heap.